File: src/MainWindow.py
from PyQt5.QtCore import Qt
from PyQt5.QtWidgets import QMainWindow, QHBoxLayout, QWidget, QPushButton, QErrorMessage
from CrossbarGrid import CrossbarGrid

"""
Class that generates the GUI and handles the button events
"""
class MainWindow(QMainWindow):
    """
    :param.title    title of the main window
    :param.model    a crossbar model instance
    """
    def __init__(self, title, model):
        super(MainWindow, self).__init__()
        self.setWindowTitle(title)
        self._model = model
        self._draw_window()

    def _draw_window(self):
        central_widget = QWidget(self)

        # Crossbar layout
        grid = CrossbarGrid(central_widget, self._model)

        # Evolve button
        evolve_button = QPushButton("Evolve", self)
        evolve_button.clicked.connect(self._safe_evolve)

        # Run code
        #run_code_button = QPushButton("Run code", self)
        #run_code_butto.clicked.connect(self._safe_run_code)

        layout = QHBoxLayout()
        layout.addWidget(grid)
        layout.addWidget(evolve_button)

        central_widget.setLayout(layout)
        self.setCentralWidget(central_widget)

    """
    Simulate the next time-step of the crossbar
    """
    def _safe_evolve(self):
        try:
            self._model.evolve()
        except Exception as e:
            print(e.args)
            error_dialog = QErrorMessage(self)
            error_dialog.setWindowModality(Qt.WindowModal)
            error_dialog.showMessage(e.args[0])
	
    """
    Simulate the QASM code in the crossbar
    """
    # TODO
    def _safe_run_code(self):
        try:
            self._model.evolve()
        except Exception as e:
            print(e.args)

File: gui/MainWindow.cpp
#include <iostream>
#include <regex>
#include <QErrorMessage>
#include "MainWindow.h"
#include "gui/modals/QubitInfo.h"

MainWindow::MainWindow(QWidget* parent) : QMainWindow(parent) {
	this->parent = parent;
	this->ui = new Ui::MainWindow();
	this->ui->setupUi(this);

	this->model = new CrossbarModel(8, 8);
	this->draw_window();
}

MainWindow::~MainWindow() {
	delete ui;
}

void MainWindow::draw_window() {
	// Menu actions
	this->connect(ui->actionOpen, SIGNAL(triggered()), this, SLOT(open_file()));
	this->connect(ui->actionSettings, SIGNAL(triggered()), this, SLOT(open_settings()));
	this->connect(ui->actionExit, SIGNAL(triggered()), this, SLOT(exit()));

	// Evolve button
	this->connect(ui->evolveButton, SIGNAL(clicked()), this, SLOT(safe_evolve()));

	// Qubit States button
	this->connect(ui->qubitInfoButton, SIGNAL(clicked()), this, SLOT(open_qubit_info()));

	// Reset button
	this->connect(ui->resetButton, SIGNAL(clicked()), this, SLOT(safe_reset()));
	
	// Validate code button
	this->connect(ui->validateCodeButton, SIGNAL(clicked()), this, SLOT(safe_validate_code()));
	
	// Run code button
	this->connect(ui->runCodeButton, SIGNAL(clicked()), this, SLOT(safe_run_code()));

	// Editor
	delete ui->editor;
	this->editor = new CodeEditor(this);
	ui->horizontalLayout_2->insertWidget(0, this->editor);

	// Visual crossbar grid
	delete ui->crossbarGrid;
	ui->crossbarGrid = new CrossbarGrid(ui->centralWidget, this->model);
	ui->horizontalLayout_2->insertWidget(1, ui->crossbarGrid);
}

void MainWindow::open_file() {
	std::cout << "OPEN FILE" << std::endl << std::flush;
}

void MainWindow::open_settings() {
	std::cout << "SETTINGS" << std::endl << std::flush;
}

void MainWindow::exit() {
	this->close();
}

void MainWindow::safe_evolve() {
	try {
		this->model->evolve();
	} catch (const std::exception& ex) {
		// Undecidable configurations
		QErrorMessage* error_dialog = new QErrorMessage(this);
		error_dialog->showMessage(ex.what());
		error_dialog->exec();
		std::cout << "[Message] Undecidable configuration found" << std::endl;
	} catch (...) {
		// Unknown error
		QErrorMessage* error_dialog = new QErrorMessage(this);
		error_dialog->showMessage("Unknown error");
		error_dialog->exec();
		std::cout << "[Error] Unknown error thrown in evolve" << std::endl;
	}
}

void MainWindow::open_qubit_info() {
	QubitInfo* qubitStates = new QubitInfo(this, this->model);
	qubitStates->exec();
}

void MainWindow::safe_reset() {
	this->model->reset();
}

void MainWindow::safe_validate_code() {
	try {
		// TODO: validate code
		
		std::cout << "VALIDATE" << std::endl << std::flush;
		
	} catch (const std::exception& ex) {
		// Undecidable configurations
		QErrorMessage* error_dialog = new QErrorMessage(this);
		error_dialog->showMessage(ex.what());
		error_dialog->exec();
		std::cout << "[Message] Invalid instruction found" << std::endl;
	} catch (...) {
		//QErrorMessage(this) error_dialog;
		//error_dialog->setWindowModality(Qt.WindowModal);
		//error_dialog->showMessage("TEST");
		std::cout << "EXCEPTION THROWN in validate_code" << '\n';
	}
}

void MainWindow::safe_run_code() {
	try {
		// Read code from editor
		QString text = this->editor->toPlainText();
		QTextStream str(&text, QIODevice::ReadOnly);
		
		
		
		std::string line;
		for (int n = 1; !str.atEnd(); n++) {
			line = str.readLine().toUtf8().constData();

			// Match operation
			Operation* operation = Matcher::match(line);
			
			if (operation == NULL) {
				throw std::runtime_error("Invalid instruction at line " + std::to_string(n));
			} else {
				if (operation->validate_constraints()) {
					operation->execute();
				}
			}
		}
	} catch (const std::exception& ex) {
		// Undecidable configurations
		QErrorMessage* error_dialog = new QErrorMessage(this);
		error_dialog->showMessage(ex.what());
		error_dialog->exec();
		std::cout << "[Message] Invalid instruction found" << std::endl;
	} catch (...) {
		//QErrorMessage(this) error_dialog;
		//error_dialog->setWindowModality(Qt.WindowModal);
		//error_dialog->showMessage("TEST");
		std::cout << "EXCEPTION THROWN in run_code" << '\n';
	}
}
